src/esp/sensor/FisheyeSensor.cpp

```diff
@@ -0,0 +1,189 @@
+    case SensorType::Color:
+      fisheyeShaderFlags_ |= gfx::FisheyeShader::Flag::ColorTexture;
+      break;
+    case SensorType::Depth:
+      fisheyeShaderFlags_ |= gfx::FisheyeShader::Flag::DepthTexture;
+      break;
+    // TODO: Semantic
+    // sensor type list is too long, have to use default
+    default:
+      CORRADE_INTERNAL_ASSERT_UNREACHABLE();
+      break;
+  }
+  // prepare a big triangle mesh to cover the screen
+  mesh_ = Mn::GL::Mesh{};
+  mesh_.setCount(3);
+}
+
+bool FisheyeSensorSpec::operator==(const FisheyeSensorSpec& a) const {
+  return VisualSensorSpec::operator==(a) &&
+         fisheyeModelType == a.fisheyeModelType &&
+         focalLength == a.focalLength &&
+         principalPointOffset == a.principalPointOffset;
+}
+
+Mn::ResourceKey FisheyeSensor::getShaderKey() {
+  return Cr::Utility::formatString(
+      FISH_EYE_SHADER_KEY_TEMPLATE,
+      static_cast<Mn::UnsignedInt>(fisheyeSensorSpec_->fisheyeModelType),
+      static_cast<gfx::FisheyeShader::Flags::UnderlyingType>(
+          fisheyeShaderFlags_));
+}
+
+bool FisheyeSensor::drawObservation(sim::Simulator& sim) {
+  if (!hasRenderTarget()) {
+    return false;
+  }
+
+  // in case the fisheye sensor resolution changed at runtime
+  {
+    auto& res = fisheyeSensorSpec_->resolution;
+    int size = res[0] < res[1] ? res[0] : res[1];
+    bool reset = cubeMap_->reset(size);
+    if (reset) {
+      cubeMapCamera_->setProjectionMatrix(size, fisheyeSensorSpec_->near,
+                                          fisheyeSensorSpec_->far);
+    }
+  }
+
+  esp::gfx::RenderCamera::Flags flags;
+  if (sim.isFrustumCullingEnabled()) {
+    flags |= gfx::RenderCamera::Flag::FrustumCulling;
+  }
+  // generate the cubemap texture
+  cubeMap_->renderToTexture(*cubeMapCamera_, sim.getActiveSceneGraph(), flags);
+
+  auto drawWith = [&](auto& shader) {
+    if (fisheyeSensorSpec_->sensorType == SensorType::Color) {
+      shader.bindColorTexture(
+          cubeMap_->getTexture(gfx::CubeMap::TextureType::Color));
+    }
+    if (fisheyeSensorSpec_->sensorType == SensorType::Depth) {
+      shader.bindDepthTexture(
+          cubeMap_->getTexture(gfx::CubeMap::TextureType::Depth));
+    }
+    renderTarget().renderEnter();
+    shader.draw(mesh_);
+    renderTarget().renderExit();
+  };
+
+  switch (fisheyeSensorSpec_->fisheyeModelType) {
+    case FisheyeSensorModelType::DoubleSphere: {
+      Magnum::Resource<gfx::FisheyeShader, gfx::DoubleSphereCameraShader>
+          shader = getShader<gfx::DoubleSphereCameraShader>();
+      auto& actualSpec =
+          static_cast<FisheyeSensorDoubleSphereSpec&>(*fisheyeSensorSpec_);
+      (*shader)
+          .setFocalLength(actualSpec.focalLength)
+          .setPrincipalPointOffset(actualSpec.principalPointOffset)
+          .setAlpha(actualSpec.alpha)
+          .setXi(actualSpec.xi);
+      drawWith(*shader);
+    } break;
+
+      // TODO:
+      // The other FisheyeSensorModelType
```

Comment:
`default: CORRADE_INTERNAL_ASSERT_UNREACHABLE();` maybe? :) just so it properly slaps you when you forget, haha